### ISSUE TITLE
Apollo Client + NextJS, getServerSideProps memory leak, InMemoryCache

### ISSUE THREAD

**Issue by aaronkchsu** [apollographql/apollo-client#9699]
Official Apollo and NextJS recommendations are about to create a new ApolloClient instance each time when the GraphQL request should be executed in case if SSR is used.

This shows good results by memory usage, memory grows for some amount and then resets with the garbage collector to the initial level.

The problem is that the initial memory usage level constantly grows and the debugger shows that leaking is caused by the "InMemoryCache" object that is attached to the ApolloClient instance as cache storage.

We tried to use the same "InMemoryCache" instance for the all new Apollo instances and tried to disable caching customizing policies in "defaultOptions", but the leak is still present.

Is it possible to turn off cache completely? Something like setting a "false" value for the "cache" option in ApolloClient initialization? Or maybe it's a known problem with a known solution and could be solved with customization of the "InMemoryCache"?

We tried numerous options, such as force cache garbage collection, eviction of the objects in the cache, etc., but nothing helped, the leak is still here.

Thank you!

```
export function createApolloClient(apolloCache) {
    return new ApolloClient({
        cache: apolloCache,
        connectToDevTools: !!isBrowser,
        link: ApolloLink.from([
            errorLink,
            authLink,
            createUploadLink({ credentials: "include", uri }),
        ]),
        ssrMode: !isBrowser,
        typeDefs,
        resolvers,
        defaultOptions: {
            watchQuery: {
                fetchPolicy: "cache-first",
                errorPolicy: "all",
            },
            query: {
                fetchPolicy: "cache-first",
                errorPolicy: "all",
            },
        },
    });
}

export function initializeApollo(initialState = {}) {
    const _apolloCache = apolloCache || createApolloCache();

    if (!apolloCache) apolloCache = _apolloCache;

    // console.log("APOLLO_CACHE", apolloCache);
    // apolloCache.evict();
    apolloCache.gc();

    const _apolloClient = apolloClient || createApolloClient(_apolloCache);

    // For SSG and SSR always create a new Apollo Client
    if (typeof window === "undefined") return _apolloClient;
    // Create the Apollo Client once in the client
    if (!apolloClient) apolloClient = _apolloClient;

    return _apolloClient;
}

export function useApollo(initialState) {
    const store = useMemo(() => initializeApollo(initialState), [initialState]);
    return store;
}
```

On page

```
export async function getServerSideProps(context) {
    const apolloClient = initializeApollo();
    const biteId =
        (context.query && context.query.id) ||
        (context.params && context.params.id);

    const realId = delQuery(biteId);

    try {
        await apolloClient.query({
            query: FETCH_BLERP,
            variables: { _id: realId },
            ssr: true,
        });
    } catch (err) {
        console.log("Failed to fetch blerp");
    }

    return {
        props: {
            _id: realId,
            initialApolloState: apolloClient.cache.extract(),
        },
    };
}
```

**Comment by Falco-Boehnke** [comment]
We have a similar issue with basically an identical setup. The response object keeps growing and growing, despite there being cached values they are always returned from the server.

**Comment by jeffminsungkim** [comment]
Having the same issue here.

**Comment by mcousillas6** [comment]
Same issue here, any updates from the team?

**Comment by bignimbus** [comment]
Hi all, we are taking a fresh look at server-side rendering for release 3.8.  See https://github.com/apollographql/apollo-client/issues/10231 for more details and please feel free to provide feedback 🙏🏻 As we get into the implementation details we'll have a better sense of how to tackle memory management in this new paradigm.  Thanks for your patience!

**Comment by aaronkchsu** [comment]
Awesome thanks for the update!!

**Comment by samueldusek** [comment]
Hi, 🙂 

Did anybody successfully solved the issue? 

We are facing the same issue and we really need to solve this. 🙁

**Comment by joekur** [comment]
I too have been investigating a memory leak in an SSR context, and found retained objects that stick around even after manual garbage collection (via chrome devtools). These are strings and objects related to variables of a graphql query (`client.query()`). I followed these up retaining tree up to `canonicalStringify`, and noticed both retainers of `WeakMap` entries, and `Trie`.

What stuck out to me was that the `Trie` was using strong `Map`s underneath. I inspected one of these objects, and saw that while `weakness` was set to true on the `Trie`, it had a `strong` field (of type `Map`).

See here:
![Screenshot_2023-10-19_at_11_12_02 AM-2](https://github.com/apollographql/apollo-client/assets/1043396/aae8e929-7bf2-442a-af5d-f83faf9232c7)
![Screenshot 2023-10-19 at 11 21 54 AM](https://github.com/apollographql/apollo-client/assets/1043396/3a615690-36ff-4778-9915-b847f4727543)

I haven't fully wrapped my head around how `canonicalStringify` is supposed to work (or `@wry/trie`), but this looks unexpected to me, and my assumption is that these strong `Map` instances are the reason they are not being GCed from the `WeakMap` and `Trie` references.

As mentioned in https://github.com/apollographql/apollo-client/issues/9699#issuecomment-1498885263, calling `cache.gc()` (on any `InMemoryCache` instance) clears this memory.

**Comment by jerelmiller** [comment]
@joekur you might be interested in https://github.com/apollographql/apollo-client/pull/11254 which did some work to address the memory overhead that `canonicalStringify` currently utilizes. That change was released in [`3.9.0-alpha.2`](https://github.com/apollographql/apollo-client/releases/tag/v3.9.0-alpha.2). Feel free to give that a shot and see if you get some better results!

**Comment by joekur** [comment]
@jerelmiller was just giving that a look. Gave it a local test, and it does indeed look to fix the issue. Nice work! 👏

**Comment by joekur** [comment]
@jerelmiller when can we expect a 3.9.0 release?

**Comment by phryneas** [comment]
At this point we are targeting a release candidate late November/early December, and a final release about one or two weeks after that unless we get feedback that delays us.

**Comment by phryneas** [comment]
Hey everyone! 
We released the beta containing our memory story this week - you can read all about it in the [announcement blog post](https://www.apollographql.com/blog/apollo-3-9-beta-feature-spotlight-the-memory-story).

We would be very grateful if you could try it out and report us your cache measurements so we can dial in the right default cache limits (more details in the blogpost) :)

**Comment by joekur** [comment]
@phryneas any updated ETA on the 3.9.0 stable release?

**Comment by phryneas** [comment]
@joekur we just shipped RC.1, and if nothing else comes up I would guess about a week.

**Comment by phryneas** [comment]
We have recently released Apollo 3.9 which I believe should have fixed this issue - so I'm going to close this.

If this issue keeps persisting in Apollo Client >= 3.9.0, please open a new issue!
